whois-update/src/main/java/net/ripe/db/whois/update/authentication/strategy/ReferenceAuthentication.java

```diff
@@ -0,0 +1,114 @@
+            AttributeType.ADMIN_C,
+            AttributeType.ZONE_C, AttributeType.TECH_C, AttributeType.AUTHOR, AttributeType.PING_HDL,
+            AttributeType.MNT_IRT,
+            AttributeType.MNT_BY, AttributeType.MNT_DOMAINS, AttributeType.MNT_ROUTES, AttributeType.MNT_LOWER,
+            AttributeType.MNT_REF, AttributeType.MNT_NFY, AttributeType.ORG);
+
+    public ReferenceAuthentication(final AuthenticationModule credentialValidators, final RpslObjectDao rpslObjectDao) {
+        this.credentialValidators = credentialValidators;
+        this.rpslObjectDao = rpslObjectDao;
+    }
+
+    @Override
+    public boolean supports(PreparedUpdate update) {
+        return REFERENCED_OBJECT_TYPES.stream().map(update::getNewValues).anyMatch(values -> !values.isEmpty());
+    }
+
+    @Override
+    public List<RpslObject> authenticate(final PreparedUpdate update, final UpdateContext updateContext) {
+        final List<Message> authenticationMessages = Lists.newArrayList();
+        final Map<RpslObject, List<RpslObject>> candidatesMap = getCandidates(update, updateContext);
+        final Set<RpslObject> authenticatedObjects = Sets.newLinkedHashSet();
+        for (final Map.Entry<RpslObject, List<RpslObject>> candidatesEntry : candidatesMap.entrySet()) {
+            final List<RpslObject> candidates = candidatesEntry.getValue();
+            final List<RpslObject> authenticatedBy = credentialValidators.authenticate(update, updateContext, candidates, getClass());
+
+            if (authenticatedBy.isEmpty()) {
+                final RpslObject candidate = candidatesEntry.getKey();
+                authenticationMessages.add(UpdateMessages.authenticationFailed(candidate, AttributeType.MNT_REF, candidates));
+            } else {
+                authenticatedObjects.addAll(authenticatedBy);
+            }
+        }
+
+        if (!authenticationMessages.isEmpty()) {
+            throw new AuthenticationFailedException(authenticationMessages, Sets.newLinkedHashSet(Iterables.concat(candidatesMap.values())));
+        }
+
+        return Lists.newArrayList(authenticatedObjects);
+    }
+
+    private List<RpslObject> getAllObjects(final Set<ObjectType> objectsTypes, final Collection<CIString> newOrgReferences) {
+        final List<RpslObject> objects = Lists.newArrayList();
+        objectsTypes.forEach( objectType -> objects.addAll(rpslObjectDao.getByKeys(objectType, newOrgReferences)));
+        return objects;
+    }
+    private boolean isSelfReference(final PreparedUpdate update, final Collection<CIString> newReferences,
+                                    final Set<ObjectType> objectTypes) {
+        return objectTypes.contains(update.getType()) && newReferences.contains(update.getUpdatedObject().getKey());
+    }
+
+    private Map<RpslObject, List<RpslObject>> getCandidates(final PreparedUpdate update, final UpdateContext updateContext) {
+        final Map<RpslObject, List<RpslObject>> candidates = new LinkedHashMap<>();
```

Comment:
Use `Maps.newHashMap()` for consistency, unless you specifically need an insertion-ordered Map